### ISSUE TITLE
Exception with ara running in AWX 

### ISSUE THREAD

**Issue by stokkie90** [ansible-community/ara#211]
## What is the issue ?

When Running ARA inside AWX i do get exceptions on some methods. As a result the playbook will never show finished in ARA and the data is incomplete. When running it from my workstation there is no problem. 

Exceptions:
- `Failure using method (v2_playbook_on_play_start) in callback plugin (<ansible.plugins.callback.ara_default.CallbackModule object at 0x7f5a4d94fda0>): 'id'`
- `Failure using method (v2_runner_on_skipped) in callback plugin (<ansible.plugins.callback.ara_default.CallbackModule object at 0x7f5a4d94fda0>): 'NoneType' object is not subscriptable`
- `Failure using method (v2_playbook_on_stats) in callback plugin
(<ansible.plugins.callback.ara_default.CallbackModule object at
0x7f5a4d94fda0>): 'id'`

Versions:

- Ansible Version: 2.9.18
- AWX Version 17.0.1
- ARA Version: 1.5.4 

Server side Logs at the end of the playbook:

```
[ara-7699dc477c-s9s9t ara] 10.200.212.83 - admin [04/Mar/2021:10:08:43 +0000] "PATCH /api/v1/plays/3 HTTP/1.1" 200 328 "-" "ara-http-client_1.5.4" 
[ara-7699dc477c-s9s9t ara] Bad Request: /api/v1/plays 
[ara-7699dc477c-s9s9t ara] 2021-03-04 10:08:44,077 WARNING django.request: Bad Request: /api/v1/plays 
[ara-7699dc477c-s9s9t ara] 10.200.212.83 - admin [04/Mar/2021:10:08:44 +0000] "POST /api/v1/plays HTTP/1.1" 400 34 "-" "ara-http-client_1.5.4" 
```
Related Client log message:

```
WARNING]: Failure using method (v2_playbook_on_stats) in callback plugin
(<ansible.plugins.callback.ara_default.CallbackModule object at
0x7f5a4d94fda0>): 'id'
Callback Exception: 
  File "/opt/custom-venvs/xx/lib/python3.6/site-packages/ansible/executor/task_queue_manager.py", line 381, in send_callback
    method(*new_args, **kwargs)
   File "/opt/custom-venvs/xx/lib/python3.6/site-packages/ara/plugins/callback/ara_default.py", line 405, in v2_playbook_on_stats
    self._end_play()
   File "/opt/custom-venvs/xx/lib/python3.6/site-packages/ara/plugins/callback/ara_default.py", line 430, in _end_play
    "/api/v1/plays/%s" % self.play["id"],    
```

**Comment by dmsimard** [comment]
Hi @stokkie90 and thanks for the issue o/

When there's an exception, it tends to cascade down -- for example, if ara can't create the playbook, it won't be able to create the play because it needs to be associated to the playbook and so on. I have a feeling this is what is happening here but can't tell for sure.

To try and help me narrow it down, could you confirm if my assumptions and understanding are correct:
- You have playbooks from AWX in ara but not their data (plays, files, tasks, hosts, results)
- You are running the API server as a service somewhere and using ``ARA_API_CLIENT=http``

Then, could you please:
- tell me what database backend the server is using
- see if there is additional insight in the output or the logs by running both the server and callback plugin with ``ARA_DEBUG=true`` and ``ARA_LOG_LEVEL=DEBUG``

Thanks !

**Comment by stokkie90** [comment]
As you can see their is quite some data being successfully send to ARA. 

![Screenshot from 2021-03-05 10-49-36](https://user-images.githubusercontent.com/10831825/110098522-98661b80-7da0-11eb-9445-4db358b1887c.png)

Database backend is currently still sqlite. Planning to move it to Mysql/Postgresql

I don't have more logging that what i posted in the first message. 

Could it somehow be related to running multiple plays in 1 file? looking at the output it's starting from the second play. The strange this is that i don't have this locally on my workstation.

**Comment by dmsimard** [comment]
@stokkie90 sqlite is prone to "[database is locked](https://github.com/ansible-community/ara/issues/195)" exceptions, especially when running multiple playbooks concurrently (or when multithreading is enabled, i.e ``ARA_CALLBACK_THREADS=4``) but it doesn't look like you are running into that according to the logs you provided.

When you try to reproduce the issue locally, are you using the default offline client or the http one ?

**Comment by stokkie90** [comment]
When I run it locally is against the same endpoint. So there is something strange going on in awx.

**Comment by dmsimard** [comment]
@stokkie90 have you been able to find out what was the issue ? If not, can you provide more details like I asked in my previous answers ? Thanks.

**Comment by jhampson-dbre** [comment]
I have similar error message when using ARA with Ansible runner in a playbook with `serial: 1` and at least 2 hosts in the play batch. I don't know that it is the same problem, but it works correctly with ansible-playbook cli and produces similar error messages as above. In ARA web UI, only tasks for one server are recorded but not the second server in the play and permanently shows in ARA is "in progress or interrupted".

I've tried setting ARA_DEBUG=true and ARA_LOG_LEVEL=DEBUG in both the plugin environment and the ARA server environment, but i don't get any additional output, so i assume i'm doing it wrong?

I can reproduce on demand with this playbook ran via ansible-runner:
```yaml
- hosts: all 
  gather_facts: yes
  serial: 1
  vars:
    simulate_failure: false
    simulate_long_job_seconds: 1

  tasks:
    - name: Simulate long running job 
      pause:
        seconds: "{{ simulate_long_job_seconds|int }}"

    - name: Simulate successful completion 
      debug:
        msg: "Everything is awesome"
```

produces this output
```
PLAY [all] *********************************************************************

TASK [Gathering Facts] *********************************************************
ok: [server01]

TASK [Simulate long running job] ***********************************************
Pausing for 1 seconds
(ctrl+C then 'C' = continue early, ctrl+C then 'A' = abort)
ok: [server01]

TASK [Simulate successful completion] ******************************************
ok: [server01] => {
    \msg\: \Everything is awesome\
}

PLAY [all] *********************************************************************
Failed to post on /api/v1/plays: {'name': 'all', 'status': 'running', 'uuid': '0a58ac14-3580-67b0-5b80-000000000073_2', 'playbook': 17, 'started': '2021-04-01T15:31:35.138694'}

TASK [Gathering Facts] *********************************************************
[WARNING]: Failure using method (v2_playbook_on_task_start) in callback plugin
(<ansible.plugins.callback.ara_default.CallbackModule object at
0x7fcfd509dc88>): 'id'
ok: [server03]

TASK [Simulate long running job] ***********************************************
Pausing for 1 seconds
(ctrl+C then 'C' = continue early, ctrl+C then 'A' = abort)
ok: [server03]

TASK [Simulate successful completion] ******************************************
ok: [server03] => {
    \msg\: \Everything is awesome\
}

PLAY RECAP *********************************************************************
server01             : ok=3    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
server03             : ok=3    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

[WARNING]: Failure using method (v2_playbook_on_stats) in callback plugin
(<ansible.plugins.callback.ara_default.CallbackModule object at
0x7fcfd509dc88>): 'id'
```

Only thing I see in ARA logging is:
```
172.20.53.128 - - [01/Apr/2021:15:31:27 +0000] "POST /api/v1/playbooks HTTP/1.1" 201 1228 "-" "ara-http-client_1.5.3"
172.20.53.128 - - [01/Apr/2021:15:31:27 +0000] "POST /api/v1/files HTTP/1.1" 201 593 "-" "ara-http-client_1.5.3"
172.20.53.128 - - [01/Apr/2021:15:31:27 +0000] "PATCH /api/v1/playbooks/17 HTTP/1.1" 200 1401 "-" "ara-http-client_1.5.3"
172.20.53.128 - - [01/Apr/2021:15:31:27 +0000] "POST /api/v1/plays HTTP/1.1" 201 250 "-" "ara-http-client_1.5.3"
172.20.53.128 - - [01/Apr/2021:15:31:27 +0000] "POST /api/v1/files HTTP/1.1" 201 295 "-" "ara-http-client_1.5.3"
172.20.53.128 - - [01/Apr/2021:15:31:27 +0000] "POST /api/v1/tasks HTTP/1.1" 201 307 "-" "ara-http-client_1.5.3"
172.20.53.128 - - [01/Apr/2021:15:31:33 +0000] "POST /api/v1/hosts HTTP/1.1" 201 196 "-" "ara-http-client_1.5.3"
172.20.53.128 - - [01/Apr/2021:15:31:33 +0000] "POST /api/v1/results HTTP/1.1" 201 2986 "-" "ara-http-client_1.5.3"
172.20.53.128 - - [01/Apr/2021:15:31:33 +0000] "PATCH /api/v1/tasks/271 HTTP/1.1" 200 347 "-" "ara-http-client_1.5.3"
172.20.53.128 - - [01/Apr/2021:15:31:33 +0000] "PATCH /api/v1/hosts/47 HTTP/1.1" 200 2809 "-" "ara-http-client_1.5.3"
172.20.53.128 - - [01/Apr/2021:15:31:34 +0000] "POST /api/v1/tasks HTTP/1.1" 201 303 "-" "ara-http-client_1.5.3"
172.20.53.128 - - [01/Apr/2021:15:31:35 +0000] "POST /api/v1/results HTTP/1.1" 201 486 "-" "ara-http-client_1.5.3"
172.20.53.128 - - [01/Apr/2021:15:31:35 +0000] "PATCH /api/v1/tasks/272 HTTP/1.1" 200 343 "-" "ara-http-client_1.5.3"
172.20.53.128 - - [01/Apr/2021:15:31:35 +0000] "POST /api/v1/tasks HTTP/1.1" 201 308 "-" "ara-http-client_1.5.3"
172.20.53.128 - - [01/Apr/2021:15:31:35 +0000] "POST /api/v1/results HTTP/1.1" 201 352 "-" "ara-http-client_1.5.3"
172.20.53.128 - - [01/Apr/2021:15:31:35 +0000] "PATCH /api/v1/tasks/273 HTTP/1.1" 200 348 "-" "ara-http-client_1.5.3"
2021-04-01 15:31:35,151 WARNING django.request: Bad Request: /api/v1/plays
172.20.53.128 - - [01/Apr/2021:15:31:35 +0000] "POST /api/v1/plays HTTP/1.1" 400 34 "-" "ara-http-client_1.5.3"
172.20.53.128 - - [01/Apr/2021:15:31:35 +0000] "PATCH /api/v1/plays/17 HTTP/1.1" 200 290 "-" "ara-http-client_1.5.3"
172.20.53.128 - - [01/Apr/2021:15:31:35 +0000] "POST /api/v1/files HTTP/1.1" 201 461 "-" "ara-http-client_1.5.3"
172.20.53.128 - - [01/Apr/2021:15:31:42 +0000] "POST /api/v1/hosts HTTP/1.1" 201 196 "-" "ara-http-client_1.5.3"
```
if you think this should be logged as a separate issue, then I can do that.

**Comment by tkwy-rick** [comment]
@jhampson-dbre Yeah it's exactly the same problem. Serial is indeed being used in those playbooks. 

```
---

- hosts: all
  become: yes
  tasks:
    - include_role:
        name: "{{ item }}"
      loop: "{{ overwrite_ansible_roles | default(tags.ansible_roles.split(',')) }}"
      tags:
        - always
      when:
        - tags.ansible_serial_roles is not defined or item not in tags.ansible_serial_roles.split(',')


- hosts: tag_ansible_serial_true
  become: yes
  serial: 1
  tasks:
    - include_role:
        name: "{{ item }}"
      loop: "{{ overwrite_ansible_roles | default(tags.ansible_serial_roles.split(',')) }}"
      tags:
        - always
      when:
        - tags.ansible_serial_roles is defined
        - item in tags.ansible_serial_roles.split(',')
```

**Comment by dmsimard** [comment]
Thanks for finding out about the issue being possibly related to serial !

@tkwy-rick are you also using the default sqlite backend like @jhampson-dbre ?

I don't have time to investigate this right now but I wonder if we would be able to reproduce this with postgresql or mysql.
Given AWX's reliance on postgresql, it might be a good candidate to test first. If you would like to find out, the configuration in ``settings.yaml`` should look like this:
```
default:
  # ...
  DATABASE_CONN_MAX_AGE: 300
  DATABASE_ENGINE: django.db.backends.postgresql
  DATABASE_NAME: ara
  DATABASE_USER: ara
  DATABASE_PASSWORD: secret
  DATABASE_PORT: 5432
```

Django needs ``psycopg2`` to be installed either from pip or from distribution packages when using postgresql. If installing from pip you need some dependencies installed first, on EL8 these are:
  - postgresql
  - postgresql-devel
  - python3-devel
  - gcc

**Comment by jhampson-dbre** [comment]
Before doing any other troubleshooting, do you have any advice for more detailed logging? Would really like to see the debug log for 
```
[ara-7699dc477c-s9s9t ara] Bad Request: /api/v1/plays
```

As I mentioned, I've tried setting ARA_DEBUG=true and ARA_LOG_LEVEL=DEBUG in both the plugin environment and the ARA server environment, but i only see INFO logs writing to stdout from django. I figure I'm not looking in the right place.

**Comment by jhampson-dbre** [comment]
It does look like ansible-runner does some extra massaging of play uuid when serial is used:
https://github.com/ansible/ansible-runner/blob/58e0f1e0107b3ed6bfc0af4e613adfffffb9bd2f/ansible_runner/display_callback/module.py#L203

My guess the django uuidfield type is not wide enough to handle the extra data appended by ansible-runner to make the uuid unique

**Comment by tkwy-rick** [comment]
@jhampson-dbre that could indeed be related. ARA is using UUIDField in the model play. https://github.com/ansible-community/ara/blob/master/ara/api/models.py#L182

That's limited to 32 char limit. so won't work indeed:

```
sqlite> .schema plays
CREATE TABLE IF NOT EXISTS "plays" ("id" integer NOT NULL PRIMARY KEY AUTOINCREMENT, "created" datetime NOT NULL, "updated" datetime NOT NULL, "started" datetime NOT NULL, "ended" datetime NULL, "name" varchar(255) NULL, "uuid" char(32) NOT NULL, "playbook_id" bigint NOT NULL REFERENCES "playbooks" ("id") DEFERRABLE INITIALLY DEFERRED, "duration" bigint NULL, "status" varchar(25) NOT NULL);
CREATE INDEX "plays_playbook_id_cc61e267" ON "plays" ("playbook_id");

```

**Comment by dmsimard** [comment]
> It does look like ansible-runner does some extra massaging of play uuid when serial is used:
https://github.com/ansible/ansible-runner/blob/58e0f1e0107b3ed6bfc0af4e613adfffffb9bd2f/ansible_runner/display_callback/module.py#L203

Wow, nice find.

> My guess is the django uuidfield type is not wide enough to handle the extra data appended by ansible-runner to make the uuid unique

It makes sense because then it's no longer a UUID.

The reason why ara stores the play UUID is to be able to identify which playbook is running /while it is running/, for example: https://github.com/ansible-community/ara/blob/52ca2898c23fe213ac2a9de348d2d8a113b77a04/ara/plugins/action/ara_playbook.py#L87-L96

From ara's perspective, I think the non-suffixed UUID is probably sufficient to uniquely identify it and we could discard/truncate if there's anything added to it. Would you like to try a patch that truncates the UUID length to see if that fixes it ? See this PR: https://github.com/ansible-community/ara/pull/250

**Comment by jhampson-dbre** [comment]
I haven't tried this yet, but I think ansible-runner does the same on task uuid as well, so if ARA is storing task uuid it would need some treatment. I'll see if I can test that theory with the patch

**Comment by tkwy-rick** [comment]
I just tested it. and it works! Playbooks succeeded without any ARA Warnings. And within ara it also finished!

**Comment by dmsimard** [comment]
@jhampson-dbre ara doesn't do anything with the task UUID, only the play UUID, so it should work like it did for @tkwy-rick.

**Comment by jhampson-dbre** [comment]
I think truncating the uuid size should be a safe fix. its just reverting it to the format originally generated by Ansible core. Hard to see how that could cause a problem.

**Comment by tkwy-rick** [comment]
Or we go for something similar like is being done in AWX: https://github.com/ansible/awx/blob/devel/awx/main/models/events.py#L225

But not sure if that will work within ARA

**Comment by jhampson-dbre** [comment]
It looks like AWX also has a process to truncate keys that are too long (instead of allowing it to cause an error): https://github.com/ansible/awx/blob/2acc488adfe815750e2c1f44e9dd1415e3a141ee/awx/main/models/events.py#L28

Seems like truncating to max length to whatever the field size is would be good to keep in ARA either way.

**Comment by dmsimard** [comment]
The fix (https://github.com/ansible-community/ara/pull/250) has merged into master and will be released in 1.5.6 soon enough along with the UI refactor.

Thanks for hunting this down !
